hepcrawl/spiders/oaipmh_spider.py

```diff
@@ -0,0 +1,196 @@
+
+    def start_requests(self):
+        self.from_date = self.from_date or self._resume_from
+        started_at = datetime.utcnow()
+
+        logger.info("Starting harvesting of {url} with set={set} and "
+                    "metadataPrefix={metadata_prefix}, from={from_date}, "
+                    "until={until_date}".format(
+            url=self.url,
+            set=self.set,
+            metadata_prefix=self.metadata_prefix,
+            from_date=self.from_date,
+            until_date=self.until_date
+        ))
+
+        request = Request('oaipmh+{}'.format(self.url), self.parse)
+        yield request
+
+        now = datetime.utcnow()
+        self._save_run(started_at)
+
+        logger.info("Harvesting completed. Next harvesting will resume from {}"
+                    .format(self.until_date or self.granularity.format(now)))
+
+    def parse_record(self, record):
+        """
+        This method need to be reimplemented in order to provide special parsing.
+
+        Args:
+            record (scrapy.selector.Selector): selector on the parsed record
+        """
+        raise NotImplementedError()
+
+    def parse(self, response):
+        sickle = Sickle(self.url, class_mapping={
+            'ListRecords': self.record_class,
+            'GetRecord': self.record_class,
+        })
+        try:
+            records = sickle.ListRecords(**{
+                'metadataPrefix': self.metadata_prefix,
+                'set': self.set,
+                'from': self.from_date,
+                'until': self.until_date,
+            })
+        except NoRecordsMatch as err:
+            logger.warning(err)
+            raise StopIteration()
+        for record in records:
+            response = XmlResponse(self.url, encoding='utf-8', body=record.raw)
+            selector = Selector(response, type='xml')
+            yield self.parse_record(selector)
+
+    def _make_alias(self):
+        return '{url}?metadataPrefix={metadata_prefix}&set={set}'.format(
+            url=self.url,
+            metadata_prefix=self.metadata_prefix,
+            set=self.set
+        )
+
+    def _last_run_file_path(self):
+        """Render a path to a file where last run information is stored.
+
+        Returns:
+            string: path to last runs path
+        """
+        lasts_run_path = self.settings['LAST_RUNS_PATH']
+        file_name = hashlib.sha1(self._make_alias()).hexdigest() + '.json'
+        return path.join(lasts_run_path, self.name, file_name)
+
+    def _load_last_run(self):
+        """Return stored last run information
+
+        Returns:
+            Optional[dict]: last run information or None if don't exist
+        """
+        file_path = self._last_run_file_path()
+        try:
+            with open(file_path) as f:
+                last_run = json.load(f)
+                logger.info('Last run file loaded: {}'.format(repr(last_run)))
+                return last_run
+        except IOError:
```

Comment:
Here check only if the file exists (first run) but any other errors should raise

Reply:
And if it's not there, raise an exception (a subclass of `Exception` with a meaningful name), then try-catch it in whomever calls this.